src/neighborsearch/filtered.jl

```diff
@@ -0,0 +1,146 @@
+# ------------------------------------------------------------------
+# Licensed under the MIT License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+    FilteredSearch(method, nmax=Inf; maxpercategory, maxpersector, order=Euclidean())
+
+A method for searching at most `nmax` neighbors using `method`. Extra
+constraints available: `maxpercategory` and `maxpersector`. The priority is
+given to the nearest neighbor using `order`.
+
+## Max per category
+
+It can be a `NamedTuple` or `Dict`, where the first element is the property name
+and the second defines the max neighbor per available categories of the property.
+In the example below, no more than 2 neighbors from the same hole are selected.
+
+`FilteredSearch(method, maxpercategory = (holeid = 2,))`
+[ASCII illustration]
+
+## Max per sector
+
+Only allow a max number of neighbors inside each sector. The sectors are the
+cartesian quadrants (2-D) or octants (3-D). If the neighborhood is an `Ellipsoid`,
+the sectors will match the rotated quadrants/octants.
+
+`FilteredSearch(method, maxpersector = 2)`
+[ASCII illustration]
+"""
+struct FilteredSearch{M<:NeighborSearchMethod} <: BoundedNeighborSearchMethod
+  method::M
+  nmax::Int
+  maxpercategory
+  maxpersector
+  order
+end
+
+FilteredSearch(method::M, nmax=0; maxpercategory=nothing,
+               maxpersector=nothing, order=Euclidean()) where {M} =
+  FilteredSearch{M}(method, nmax, maxpercategory, maxpersector, order)
+
+maxneighbors(method::FilteredSearch) = method.nmax
+
+function search!(neighbors, pₒ::Point, method::FilteredSearch; mask=nothing)
+  meth = method.method
+  obj  = meth.domain
+  inds = search(pₒ, meth, mask=mask)
+  nmax = method.nmax == 0 ? Inf : method.nmax
+
+  # get distances and give priority to closest neighbors if necessary
+  if !(meth isa KNearestSearch)
```

Comment:
Maybe we need to introduce a new subtype of SearchMethod? We are just trying to catch all search methods that are based on balls?

Reply:
This is also new. The idea is to sort data according to distance, giving priority to closest points when filtering data. I'm checking here, currently `KNearestSearch` (and also `KBallSearch` that I forgot here) return the `inds` sorted by distance. In a `Ellipsoid` for example, the output of `FilteredSearch` will not guarantee that the priority will be given to the closest points

Reply:
Solutions to it: (1) set `sortres=true` in `NearestNeighbors.inrange` function too. (2) sort during `FilteredSearch`.  I'd say that's better to sort during `FilteredSearch` because the filter priority could be given to a different metric than the first seacrh. e.g. an Ellipsoid would give priority to the closest "anisotropic distance", while in a filtered search with order=Euclidean() would give priority to the closest euclidean dist.

Reply:
`if !(meth isa KNearestSearch)` was an attempt to avoid sorting if it was already sorted. But this might not be enough, since `KNearestSearch` and `FilteredSearch` metrics might not be the same. Moreover, `KBallSearch` would need to be considered too as it's already sorted with `sortres=true` too

Reply:
I am trying to grasp the bits, but I get that this if statement was introduce to sort the indices returned by the inner search method differently. I think a better design would provide a sorting method in the `FilteredSearch` if that is important. I would start very simple with the options you originally considered, possibly sharing examples in the PR so that we understand what each section of the code is doing.

My feeling is that we can improve the design to accommodate the filters proposed and to make the code more readable for future contributors.

Reply:
I'll change it to make it simpler. Might sort twice the data, but I think it's more important to sort it initially to avoid inconsistencies. If the user wants to preserve original order and prevent sorting, just need to set `order=nothing`. Do you think it's better to change the name of `order` to `metric` to match other functions style?

Reply:
Yes, if `order` is a distance from Distances.jl than it should be named a `metric`.